src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -321,6 +328,206 @@ public async Task<GetTransactionModel> GetTransactionAsync(string txid)
+
+            if (address != null)
+                return new BitcoinPubKeyAddress(new KeyId(address), this.Network);
+
+            return null;
+        }
+
+        /// <summary>
+        /// This will check the wallet's list of <see cref="HdAccount.InternalAddress"/>es to see if this address is
+        /// an address that received change.
+        /// </summary>
+        /// <param name="wallet">The wallet to use.</param>
+        /// <param name="coinType">The coin type associated with the wallet; Stratis or Bitcoin.</param>
+        /// <param name="txOutAddressBase58">The base58 address to verify from the <see cref="TxOut"/>.</param>
+        /// <returns><c>true</c> if the <paramref name="txOutAddressBase58"/> is a change address.</returns>
+        private bool IsChange(Wallet wallet, CoinType coinType, BitcoinPubKeyAddress txOutAddressBase58)
+        {
+            var internalAddresses = wallet.GetAccountsByCoinType(coinType).SelectMany(a => a.InternalAddresses);
+            var result = internalAddresses.FirstOrDefault(ia => ia.Address == txOutAddressBase58.ToString());
+            return result != null;
+        }
+
+        private async Task<bool> IsTxInMineAsync(Wallet wallet, CoinType coinType, IEnumerable<TransactionData> transactions, TxIn txIn)
+        {
+            var previousTransaction = transactions.FirstOrDefault(t => t.Id == txIn.PrevOut.Hash);
+            if (previousTransaction != null)
+            {
+                var prevTx = await this.blockStore.GetTransactionByIdAsync(previousTransaction.Id);
+                if (txIn.PrevOut.N < prevTx.Outputs.Count)
+                {
+                    // At this point we need to check if the scriptPubkey is in our wallet.
+                    // see: https://github.com/bitcoin/bitcoin/blob/011c39c2969420d7ca8b40fbf6f3364fe72da2d0/src/script/ismine.cpp
+
+                    var txOutScriptPubKey = prevTx.Outputs[txIn.PrevOut.N].ScriptPubKey;
+
+                    return IsAddressMine(wallet, txOutScriptPubKey, coinType);
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines whether the script translates to an address that exists in the given wallet.
+        /// </summary>
+        /// <param name="wallet">The wallet to use.</param>
+        /// <param name="scriptPubKey">The script to check.</param>
+        /// <param name="coinType">The coin type associated with the wallet; Stratis or Bitcoin.</param>
+        /// <returns><c>true</c> if the <paramref name="scriptPubKey"/> is an address in the given wallet.</returns>
+        private bool IsAddressMine(Wallet wallet, Script scriptPubKey, CoinType coinType)
+        {
+            var addressBase58 = GetBase58AddressFromScript(scriptPubKey);
+            var result = wallet.GetAllAddressesByCoinType(coinType).FirstOrDefault(a => a.Address == addressBase58.ToString());
```

Comment:
I also think that this could be retrieved once and passed in as a dictionary to be used for lookups, rather than getting multiply interrogated

Reply:
I think I will just create a variable to hold all the addresses and pass that in.